---
title: "Differential analysis and pathway enrichment"
author: "Maria Tsalenchuk"
date: "2024-10-21"
output: html_document
---

#load libraries
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(edgeR)
library(dplyr)
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(clusterProfiler)
library(stringi)
library(readr)
library(DESeq2)
library(ChIPseeker)
library(gtools)
library(patchwork)
library(png)
library(org.Rn.eg.db)
library(ggpubr)
library(GenomicRanges)
library(TxDb.Rnorvegicus.UCSC.rn7.refGene)
library(openxlsx)
library(gridExtra)
library(ggseqlogo)
library(scales)
```

## I. Differential Acetylation and Expression Analysis 

```{r ChIP import data}
mapped_reads <- read.csv("../data/bulk_rat_chip_reads.csv", header = T)
load(file = "../data/CTX_H3K27ac_peaks_counts.rda")

group <- mapped_reads$name
design <- model.matrix(~group)
keep <- filterByExpr(counts, design)
counts <- counts[keep,]
CTX_peaks <- peaks
CTX_counts <- counts

rm(counts, peaks)

# import H3K27ac peaks and counts
load("../data/SN_H3K27ac_peaks_counts.rda")
group <- mapped_reads$name
design <- model.matrix(~group)
keep <- filterByExpr(counts, design)
counts <- counts[keep,]
SN_peaks <- peaks
SN_counts <- counts
rm(counts, peaks)

SN_peaks <- SN_peaks[SN_peaks$Peak %in% row.names(SN_counts),]
```

```{r CTX H3K27ac differential analysis}
## Differential analysis for data in the cortex  ##

#IGF127225=ROT IGF127226=ROT IGF127259=CON IGF127264=CON IGF127247=CON IGF127286=CON IGF127285=ROT IGF127282=ROT IGF127287=ROT

# Define groups for each sample and set control as reference
CTXgroup = factor(c(rep("rotenone", 2), rep("control", 3), rep("rotenone", 2),"control","rotenone"))
CTXgroup <- relevel(CTXgroup, "control")

# Create DGEList object and calculate normalisation factors
CTXpeakList <- DGEList(counts = CTX_counts, group = CTXgroup)
CTXpeak_norm <- calcNormFactors(CTXpeakList)

# Quasi-likelihood F test design matrix
CTXdesign <- model.matrix(~ CTXgroup)
colnames(CTXdesign)[2] <- "Rotenone"

# Estimate dispersion and fit model
CTXDGE_norm_disp <- estimateDisp(CTXpeak_norm, CTXdesign)
CTXfit <- glmQLFit(CTXDGE_norm_disp, CTXdesign)

# Perform quasi-likelihood F test
CTXqlf <- glmQLFTest(CTXfit)

# Extract DARs
CTX_DAR <- CTXqlf$table

# Order peaks by significance
CTX_DAR$FDR <- p.adjust(CTX_DAR$PValue, method="fdr")
CTX_DAR <- CTX_DAR[order(CTX_DAR$PValue),]

# Cleanup
rm(CTXpeakList, CTXDGE_norm_disp, CTXfit, CTXqlf, CTXpeak_norm, CTXdesign)

# Gene annotation
CTX_annot <- as.data.frame(annotatePeak(makeGRangesFromDataFrame(CTX_peaks[,1:3]), 
                                            TxDb = TxDb.Rnorvegicus.UCSC.rn7.refGene,
                                            tssRegion=c(-3000, 3000), annoDb="org.Rn.eg.db"))
# Merge annotation with dataset
row.names(CTX_annot) <- CTX_peaks$Peak
CTX_annot$ID <- paste(CTX_annot$start, sep=".", CTX_annot$end)
CTX_peaks$ID <- paste(CTX_peaks$Start, sep=".", CTX_peaks$End)
m <- CTX_peaks$ID %in% CTX_annot$ID
CTX_peaks <- CTX_peaks[m,]
row.names(CTX_annot) <- CTX_peaks$Peak

# summary table: add DAR category and sort by p value
CTX_DAR_annot <- merge(CTX_DAR, CTX_annot[,c(1:3,6,14,16,17)], by = "row.names")
CTX_DAR_annot <- CTX_DAR_annot[,c(7,8,9,10,2,5,6,12,13,11)]
names(CTX_DAR_annot)[c(1,8)] <-  c("chr","GENE")
CTX_DAR_annot <- CTX_DAR_annot %>% mutate(DAR = ifelse(logFC > 0 & FDR < 0.05, "Hyper",
              ifelse(logFC < 0 & FDR < 0.05, "Hypo", "non-DAR"))) %>% arrange(PValue)
CTX_DAR_annot$DAR <- factor(CTX_DAR_annot$DAR, levels=c("Hyper", "Hypo", "non-DAR"))
```
 
```{r SN H3K27ac differential analysis}
## Differential analysis for H3K27ac data in the substantia nigra (ventral midbrain)  ##

# Define groups for each sample and set control as reference
SNgroup = factor(c(rep("rotenone",5),rep("control",4)))
SNgroup <- relevel(SNgroup, "control")

# Create DGEList object and calculate normalisation factors
SNpeakList <- DGEList(counts = SN_counts, group = SNgroup)
SNpeak_norm <- calcNormFactors(SNpeakList)
rm()

# Quasi-likelihood F test design matrix
SNdesign <- model.matrix(~ SNgroup)
colnames(SNdesign)[2] <- "Rotenone"

# Estimate dispersion and fit model
SNDGE_norm_disp <- estimateDisp(SNpeak_norm, SNdesign)
SNfit <- glmQLFit(SNDGE_norm_disp, SNdesign)

# Perform quasi-likelihood F test
SNqlf <- glmQLFTest(SNfit)

# Extract DARs
SN_DAR <- SNqlf$table

# Order peaks by significance
SN_DAR$FDR <- p.adjust(SN_DAR$PValue, method="fdr")
SN_DAR <- SN_DAR[order(SN_DAR$PValue),]

# Cleanup
rm(SNDGE_norm_disp, SNfit, SNqlf, SNpeak_norm, SNdesign,SNpeakList)

# Gene annotation
SN_annot <- as.data.frame(annotatePeak(makeGRangesFromDataFrame(SN_peaks[,1:3]), 
                                            TxDb = TxDb.Rnorvegicus.UCSC.rn7.refGene,
                                            tssRegion=c(-3000, 3000), annoDb="org.Rn.eg.db"))
# Merge annotation with dataset
SN_annot$ID <- paste(SN_annot$start, sep=".", SN_annot$end)
SN_peaks$ID <- paste(SN_peaks$Start, sep=".", SN_peaks$End)
m <- SN_peaks$ID %in% SN_annot$ID
SN_peaks <- SN_peaks[m,]
row.names(SN_annot) <- SN_peaks$Peak

# summary table: add DAR category and sort by p value
SN_DAR_annot <- merge(SN_DAR, SN_annot[,c(1:3,6,14,16,17)], by = "row.names")
SN_DAR_annot <- SN_DAR_annot[,c(7,8,9,10,2,5,6,12,13,11)]
names(SN_DAR_annot)[c(1,8)] <-  c("chr","GENE")
SN_DAR_annot <- SN_DAR_annot %>% mutate(DAR = ifelse(logFC > 0 & FDR < 0.05, "Hyper",
              ifelse(logFC < 0 & FDR < 0.05, "Hypo", "non-DAR"))) %>% arrange(PValue)
SN_DAR_annot$DAR <- factor(SN_DAR_annot$DAR, levels=c("Hyper", "Hypo", "non-DAR"))
```

```{r supplementary table significant DAR}
CTX_hyper_DAR <- CTX_DAR_annot %>% filter(DAR == "Hyper")
CTX_hypo_DAR <- CTX_DAR_annot %>% filter(DAR == "Hypo")

write.xlsx(CTX_hyper_DAR, file = "../tables/CTX_Hyper_DARs.xlsx", row.names = FALSE)
write.xlsx(CTX_hypo_DAR, file = "../tables/CTX_Hypo_DARs.xlsx", row.names = FALSE)

SN_hyper_DAR <- SN_DAR_annot %>% filter(DAR == "Hyper")
SN_hypo_DAR <- SN_DAR_annot %>% filter(DAR == "Hypo")

write.xlsx(SN_hyper_DAR, file = "../tables/SN_Hyper_DARs.xlsx", row.names = FALSE)
write.xlsx(SN_hypo_DAR, file = "../tables/SN_Hypo_DARs.xlsx", row.names = FALSE)

rm(SN_hyper_DAR, SN_hypo_DAR, CTX_hyper_DAR, CTX_hypo_DAR)
```

```{r RNA import data}
CTX_counts_rna <- read.csv("../data/CTX_counts_rna.csv", row.names = 1, check.names = FALSE)
SN_counts_rna <- read.csv("../data/SN_counts_rna.csv", row.names = 1, check.names = FALSE)
metadata <- read.csv("../data/rna_metadata.csv")
```

```{r differential gene expression CTX}
# Create DGEList and calculate normalization factors
DGE <- DGEList(counts = CTX_counts_rna, group = metadata[metadata$ngsid %in% names(CTX_counts_rna),]$treatment)
DGE_norm <- calcNormFactors(DGE)

# Quasi-likelihood F test
design <- model.matrix(~ metadata[metadata$ngsid %in% names(CTX_counts_rna),]$treatment)
colnames(design)[2] <- "Rotenone"
DGE_norm_disp <- estimateDisp(DGE_norm, design)
fit <- glmQLFit(DGE_norm_disp, design)
qlf <- glmQLFTest(fit)
DGE_CTX <- qlf$table
rm(DGE, DGE_norm, design, DGE_norm_disp, fit, qlf)

# DGE summary table
DGE_CTX$FDR <- p.adjust(DGE_CTX$PValue, method="fdr")
DGE_CTX <- DGE_CTX[order(DGE_CTX$PValue),]
DGE_CTX$Gene <- row.names(DGE_CTX)
DGE_CTX <- DGE_CTX %>% mutate(DEG = ifelse(logFC > 0 & FDR < 0.05, "UP",
                  ifelse(logFC < 0 & FDR < 0.05, "DOWN", "non-DEG")))
DGE_CTX$DEG <- factor(DGE_CTX$DEG, levels=c("UP", "DOWN", "non-DEG"))
```

```{r differential gene expression SNpc}
# Create DGEList and calculate normalization factors
DGE <- DGEList(counts = SN_counts_rna, group = metadata[metadata$ngsid %in% names(SN_counts_rna),]$treatment)
DGE_norm <- calcNormFactors(DGE)

# Quasi-likelihood F test
design <- model.matrix(~ metadata[metadata$ngsid %in% names(SN_counts_rna),]$treatment)
colnames(design)[2] <- "Rotenone"
DGE_norm_disp <- estimateDisp(DGE_norm, design)
fit <- glmQLFit(DGE_norm_disp, design)
qlf <- glmQLFTest(fit)
DGE_SN <- qlf$table
rm(DGE, DGE_norm, design, DGE_norm_disp, fit, qlf)

# DGE summary table
DGE_SN$FDR <- p.adjust(DGE_SN$PValue, method="fdr")
DGE_SN <- DGE_SN[order(DGE_SN$PValue),]
DGE_SN$Gene <- row.names(DGE_SN)
DGE_SN <- DGE_SN %>% mutate(DEG = ifelse(logFC > 0 & FDR < 0.05, "UP",
              ifelse(logFC < 0 & FDR < 0.05, "DOWN", "non-DEG")))
DGE_SN$DEG <- factor(DGE_SN$DEG, levels=c("UP", "DOWN", "non-DEG"))
```

```{r supplementary table significant DEG}
CTX_UP_DEG <- DGE_CTX %>% filter(DEG == "UP")
CTX_DOWN_DEG <- DGE_CTX %>% filter(DEG == "DOWN")

write.xlsx(CTX_UP_DEG, file = "../tables/CTX_UP_DEGs.xlsx", row.names = FALSE)
write.xlsx(CTX_DOWN_DEG, file = "../tables/CTX_DOWN_DEGs.xlsx", row.names = FALSE)

SN_UP_DEG <- DGE_SN %>% filter(DEG == "UP")
SN_DOWN_DEG <- DGE_SN %>% filter(DEG == "DOWN")

write.xlsx(SN_UP_DEG, file = "../tables/SN_UP_DEGs.xlsx", row.names = FALSE)
write.xlsx(SN_DOWN_DEG, file = "../tables/SN_DOWN_DEGs.xlsx", row.names = FALSE)

rm(CTX_UP_DEG,CTX_DOWN_DEG,SN_UP_DEG,SN_DOWN_DEG)
```

# Figure 1

```{r stacked barplot of DARs}
# Subset by significant values
dar_ctx <- subset(CTX_DAR_annot, FDR < 0.05)
dar_sn <- subset(SN_DAR_annot, FDR < 0.05)
dar_ctx$region <- "Cortex"
dar_sn$region <- "SN"
both <- rbind(dar_sn,dar_ctx)

# Extract counts
DAR_counts <- both %>%
  group_by(region, DAR) %>%
  summarise(count =n(), .groups ='drop')

DAR_counts$region <- factor(DAR_counts$region, levels = c("SN", "Cortex"))

stacked <- ggplot(DAR_counts, aes(x = region, y = count, fill = DAR)) +
  geom_bar(stat = "identity", position = "stack") +
  theme_classic(base_size = 18) + 
  labs(x = "Brain region", y = "Number of DARs", fill = NULL, 
       title = NULL) +
  scale_fill_manual(values = c("Hyper" = "#fadf93", "Hypo" = "#f7c845")) +
  theme(axis.title = element_text(size = 26),
        axis.text = element_text(size = 24), 
        legend.text = element_text(size = 22), 
        plot.title = element_text(size = 2, hjust = 0.5)) 
```

```{r volcano plots for H3K27ac cortex and SNpc}
## Volcano plot of H3K27ac fold change in rotenone vs control group ##
# get top 20 genes for labelling 
top_genes <- head(CTX_DAR_annot, 20)

# need to add labels for the top 20 genes
CTX_volcano <- ggplot(CTX_DAR_annot, 
                         aes(x=logFC, y=-log10(FDR), colour=DAR)) + 
  geom_point(size=1) + 
  xlim(-3,3) + 
  scale_colour_manual(values=c("#de9483", "#de9483", "#e0b1a6")) + 
  theme_classic() + 
  ylab("-log10(FDR)") + 
  guides(colour="none") + 
  theme(text = element_text(size=24), plot.title = element_text(hjust = 0.5)) + 
  ggtitle("Cortex") + 
  geom_text_repel(data = top_genes, aes(label = GENE), box.padding = 0.5, colour="#506777FF", size=5) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "slategrey") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "slategrey")

## Volcano plot of H3K27ac fold change in rotenone vs control group ##
# get top 20 genes for labelling 
top_genes <- head(SN_DAR_annot, 20)

# need to add labels for the top 20 genes
SN_volcano <- ggplot(SN_DAR_annot, 
                         aes(x=logFC, y=-log10(FDR), colour=DAR)) + 
  geom_point(size=1) + 
  xlim(-3,3) + 
  scale_colour_manual(values=c("#83d4b9", "#83d4b9", "#afe3d2")) + 
  theme_classic() + 
  ylab("-log10(FDR)") + 
  guides(colour="none") + 
  theme(text = element_text(size=24), plot.title = element_text(hjust = 0.5)) + 
  ggtitle("SN") + 
  geom_text_repel(data = top_genes, aes(label = GENE), box.padding = 0.5, colour="#506777FF", size=5) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "slategrey") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "slategrey")
```

## II. Pathway Enrichment Analysis

```{r load DGE/DAR}
load("../data/diffex.rda")
```

```{r CTX H3K27ac GO enrichment}
## GO enrichment for (1) hyperacetylated and (2) hypoacetylated regions ##

# create DAR gene list
CTX_DAR_list <- list(Hyper=CTX_DAR_annot[CTX_DAR_annot$DAR=="Hyper",]$GENE, 
                 Hypo=CTX_DAR_annot[CTX_DAR_annot$DAR=="Hypo",]$GENE)

CTX_GO_DAR <- list() 
for (x in c("Hyper", "Hypo")){
  print(x)
  
  print("GO_Molecular_Function")
  GO <- enrichGO(CTX_DAR_list[[x]],
                    OrgDb = org.Rn.eg.db, keyType = "SYMBOL", 
                    ont = "MF", pvalueCutoff = 0.05, pAdjustMethod = "fdr", 
                    universe = NULL, qvalueCutoff = 0.05, 
                    minGSSize = 10, maxGSSize = 500, readable = FALSE, pool = FALSE)
  if (nrow(GO@result) > 0) {
    CTX_GO_DAR[[x]]$GO_Molecular_Function <- as.data.frame(GO@result)
    CTX_GO_DAR[[x]]$GO_Molecular_Function$direction <- x
    CTX_GO_DAR[[x]]$GO_Molecular_Function$category <- "Molecular function"
  }
  
  print("GO_Biological_Process")
  GO <- enrichGO(CTX_DAR_list[[x]],
                    OrgDb = org.Rn.eg.db, keyType = "SYMBOL", 
                    ont = "BP", pvalueCutoff = 0.05, pAdjustMethod = "fdr", 
                    universe = NULL, qvalueCutoff = 0.05, 
                    minGSSize = 10, maxGSSize = 500, readable = FALSE, pool = FALSE)
  if (nrow(GO@result) > 0) {
    CTX_GO_DAR[[x]]$GO_Biological_Process <- as.data.frame(GO@result)
    CTX_GO_DAR[[x]]$GO_Biological_Process$direction <- x
    CTX_GO_DAR[[x]]$GO_Biological_Process$category <- "Biological process"
  }
}

# GO summary tables: all / top 10
CTX_GO_DAR_all <- do.call(rbind, lapply(unlist(CTX_GO_DAR, recursive = FALSE), as.data.frame))
CTX_GO_DAR_top <- CTX_GO_DAR_all %>% arrange(p.adjust) %>% group_by(direction) %>% slice_head(n=10)
CTX_GO_DAR_top$GeneRatioNumeric <- sapply(CTX_GO_DAR_top$GeneRatio, function(x) eval(parse(text = x)))
```

```{r SN H3K27ac GO enrichment}
## GO enrichment for (1) hyperacetylated and (2) hypoacetylated regions ##

# create DAR gene list
SN_DAR_list <- list(Hyper=SN_DAR_annot[SN_DAR_annot$DAR=="Hyper",]$GENE, 
                 Hypo=SN_DAR_annot[SN_DAR_annot$DAR=="Hypo",]$GENE)

# GO enrichment, using all peaks as background
SN_GO_DAR <- list() 
for (x in c("Hyper", "Hypo")){
  print(x)
  
  print("GO_Molecular_Function")
  GO <- enrichGO(SN_DAR_list[[x]],
                    OrgDb = org.Rn.eg.db, keyType = "SYMBOL", 
                    ont = "MF", pvalueCutoff = 0.05, pAdjustMethod = "fdr", 
                    universe = NULL, qvalueCutoff = 0.05, 
                    minGSSize = 10, maxGSSize = 500, readable = FALSE, pool = FALSE)
  if (nrow(GO@result) > 0) {
    SN_GO_DAR[[x]]$GO_Molecular_Function <- as.data.frame(GO@result)
    SN_GO_DAR[[x]]$GO_Molecular_Function$direction <- x
    SN_GO_DAR[[x]]$GO_Molecular_Function$category <- "Molecular function"
  }
  
  print("GO_Biological_Process")
  GO <- enrichGO(SN_DAR_list[[x]],
                    OrgDb = org.Rn.eg.db, keyType = "SYMBOL", 
                    ont = "BP", pvalueCutoff = 0.05, pAdjustMethod = "fdr", 
                    universe = NULL, qvalueCutoff = 0.05, 
                    minGSSize = 10, maxGSSize = 500, readable = FALSE, pool = FALSE)
  if (nrow(GO@result) > 0) {
    SN_GO_DAR[[x]]$GO_Biological_Process <- as.data.frame(GO@result)
    SN_GO_DAR[[x]]$GO_Biological_Process$direction <- x
    SN_GO_DAR[[x]]$GO_Biological_Process$category <- "Biological process"
  }
}

# GO summary tables: all / top 10
SN_GO_DAR_all <- do.call(rbind, lapply(unlist(SN_GO_DAR, recursive = FALSE), as.data.frame))
SN_GO_DAR_top <- SN_GO_DAR_all %>% arrange(p.adjust) %>% group_by(direction) %>% slice_head(n=10)
SN_GO_DAR_top$GeneRatioNumeric <- sapply(SN_GO_DAR_top$GeneRatio, function(x) eval(parse(text = x)))
```

```{r supplementary table GO chip}
CTX_GO_DAR_Hyper <- CTX_GO_DAR_all %>% filter(direction == "Hyper" & p.adjust < 0.05)
CTX_GO_DAR_Hypo <- CTX_GO_DAR_all %>% filter(direction == "Hypo" & p.adjust < 0.05)

write.xlsx(CTX_GO_DAR_Hyper, file = "../tables/CTX_GO_DAR_Hyper.xlsx", row.names = FALSE)
write.xlsx(CTX_GO_DAR_Hypo, file = "../tables/CTX_GO_DAR_Hypo.xlsx", row.names = FALSE)

SN_GO_DAR_Hyper <- SN_GO_DAR_all %>% filter(direction == "Hyper" & p.adjust < 0.05)
SN_GO_DAR_Hypo <- SN_GO_DAR_all %>% filter(direction == "Hypo" & p.adjust < 0.05)

write.xlsx(SN_GO_DAR_Hyper, file = "../tables/SN_GO_DAR_Hyper.xlsx", row.names = FALSE)
write.xlsx(SN_GO_DAR_Hypo, file = "../tables/SN_GO_DAR_Hypo.xlsx", row.names = FALSE)
```

```{r CTX GO RNA enrichment}
## GO enrichment for (1) upregulated and (2) downregulated genes in the cortex ##

# prepare DEG lists (up/down-regulated)
CTX_DEG_list <- list(UP=DGE_CTX[DGE_CTX$DEG=="UP",]$Gene, 
                 DOWN=DGE_CTX[DGE_CTX$DEG=="DOWN",]$Gene)

CTX_GO_DEG <- list() 
for (x in c("UP", "DOWN")){
  print(x)
  
  print("GO_Molecular_Function")
  GO <- enrichGO(CTX_DEG_list[[x]],
                    OrgDb = org.Rn.eg.db, keyType = "SYMBOL", 
                    ont = "MF", pvalueCutoff = 0.05, pAdjustMethod = "fdr", 
                    universe = NULL, qvalueCutoff = 0.05, 
                    minGSSize = 10, maxGSSize = 500, readable = FALSE, pool = FALSE)
  if (nrow(GO@result) > 0) {
    CTX_GO_DEG[[x]]$GO_Molecular_Function <- as.data.frame(GO@result)
    CTX_GO_DEG[[x]]$GO_Molecular_Function$direction <- x
    CTX_GO_DEG[[x]]$GO_Molecular_Function$category <- "Molecular function"
  }
  
  print("GO_Biological_Process")
  GO <- enrichGO(CTX_DEG_list[[x]],
                    OrgDb = org.Rn.eg.db, keyType = "SYMBOL", 
                    ont = "BP", pvalueCutoff = 0.05, pAdjustMethod = "fdr", 
                    universe = NULL, qvalueCutoff = 0.05, 
                    minGSSize = 10, maxGSSize = 500, readable = FALSE, pool = FALSE)
  if (nrow(GO@result) > 0) {
    CTX_GO_DEG[[x]]$GO_Biological_Process <- as.data.frame(GO@result)
    CTX_GO_DEG[[x]]$GO_Biological_Process$direction <- x
    CTX_GO_DEG[[x]]$GO_Biological_Process$category <- "Biological process"
  }
}

# GO summary tables: all / top 10
CTX_GO_DEG_all <- do.call(rbind, lapply(unlist(CTX_GO_DEG, recursive = FALSE), as.data.frame))
CTX_GO_DEG_top <- CTX_GO_DEG_all %>% arrange(p.adjust) %>% group_by(direction) %>% slice_head(n=5)
CTX_GO_DEG_top$GeneRatioNumeric <- sapply(CTX_GO_DEG_top$GeneRatio, function(x) eval(parse(text = x)))
```

```{r SN GO RNA enrichment}
## GO enrichment analysis for (1) upregulated and (2) downregulated genes in VM ##

# prepare DEG lists (up/down-regulated)
SN_DEG_list <- list(Up=DGE_SN[DGE_SN$DEG=="UP",]$Gene, 
                 Down=DGE_SN[DGE_SN$DEG=="DOWN",]$Gene)

SN_GO_DEG <- list() 
for (x in c("Up", "Down")){
  print(x)
  
  print("GO_Molecular_Function")
  GO <- enrichGO(SN_DEG_list[[x]],
                    OrgDb = org.Rn.eg.db, keyType = "SYMBOL", 
                    ont = "MF", pvalueCutoff = 0.05, pAdjustMethod = "fdr", 
                    universe = NULL, qvalueCutoff = 0.05, 
                    minGSSize = 10, maxGSSize = 500, readable = FALSE, pool = FALSE)
  if (nrow(GO@result) > 0) {
    SN_GO_DEG[[x]]$GO_Molecular_Function <- as.data.frame(GO@result)
    SN_GO_DEG[[x]]$GO_Molecular_Function$direction <- x
    SN_GO_DEG[[x]]$GO_Molecular_Function$category <- "Molecular function"
  }
  
  print("GO_Biological_Process")
  GO <- enrichGO(SN_DEG_list[[x]],
                    OrgDb = org.Rn.eg.db, keyType = "SYMBOL", 
                    ont = "BP", pvalueCutoff = 0.05, pAdjustMethod = "fdr", 
                    universe = NULL, qvalueCutoff = 0.05, 
                    minGSSize = 10, maxGSSize = 500, readable = FALSE, pool = FALSE)
  if (nrow(GO@result) > 0) {
    SN_GO_DEG[[x]]$GO_Biological_Process <- as.data.frame(GO@result)
    SN_GO_DEG[[x]]$GO_Biological_Process$direction <- x
    SN_GO_DEG[[x]]$GO_Biological_Process$category <- "Biological process"
  }
}

# GO summary tables: all / top 10
SN_GO_DEG_all <- do.call(rbind, lapply(unlist(SN_GO_DEG, recursive = FALSE), as.data.frame))
SN_GO_DEG_top <- SN_GO_DEG_all %>% arrange(p.adjust) %>% group_by(direction) %>% slice_head(n=5)
SN_GO_DEG_top$GeneRatioNumeric <- sapply(SN_GO_DEG_top$GeneRatio, function(x) eval(parse(text = x)))
```

```{r supplementary table GO RNA}
CTX_GO_DEG_Hyper <- CTX_GO_DEG_all %>% filter(direction == "UP" & p.adjust < 0.05)
CTX_GO_DEG_Hypo <- CTX_GO_DEG_all %>% filter(direction == "DOWN" & p.adjust < 0.05)

write.xlsx(CTX_GO_DEG_Hyper, file = "../tables/CTX_GO_DEG_Hyper.xlsx", row.names = FALSE)
write.xlsx(CTX_GO_DEG_Hypo, file = "../tables/CTX_GO_DEG_Hypo.xlsx", row.names = FALSE)

SN_GO_DEG_Hyper <- SN_GO_DEG_all %>% filter(direction == "Up" & p.adjust < 0.05)
SN_GO_DEG_Hypo <- SN_GO_DEG_all %>% filter(direction == "Down" & p.adjust < 0.05)

write.xlsx(SN_GO_DEG_Hyper, file = "../tables/SN_GO_DEG_Hyper.xlsx", row.names = FALSE)
write.xlsx(SN_GO_DEG_Hypo, file = "../tables/SN_GO_DEG_Hypo.xlsx", row.names = FALSE)
```

# Figure 2

```{r SN compare ChIP GO to RNA GO}
#filter DEG GO by the top 10 DAR GO
SN_GO_compare <- SN_GO_DEG_all %>%
  filter(ID %in% SN_GO_DAR_top$ID)

SN_GO_compare$GeneRatioNumeric <- sapply(SN_GO_compare$GeneRatio, function(x) eval(parse(text = x)))

#filter by most significant 
SN_GO_compare <- SN_GO_compare %>%
  group_by(ID) %>%
  slice_min(order_by = p.adjust, n = 1) %>%
  ungroup()

SN_GO_compare$assay <- "RNA"
SN_GO_DAR_top$assay <- "ChIP"

SN_GO <- rbind(SN_GO_DAR_top,SN_GO_compare)

SN_GO <- SN_GO[,c(2,6,10,12,13)]

SN_GO <- SN_GO %>%
  mutate(direction = case_when(
    direction == "Hyper" ~ "Up",
    direction == "Hypo" ~ "Down",
    TRUE ~ direction))

SN_GO_up <- SN_GO[SN_GO$direction =="Up",]
SN_GO_down <- SN_GO[SN_GO$direction =="Down",]

SN_GO_up_plot <- ggplot(SN_GO_up,
       aes(x = GeneRatioNumeric, y = reorder(Description, -log10(p.adjust)))) +
  geom_point(aes(size = -log10(p.adjust), color = -log10(p.adjust)), alpha = 0.7) + 
  scale_color_gradient(low = "#f0c560", high = "#e37c07") + 
  scale_size(range = c(1, 10)) + 
  theme_minimal() + 
  theme(
    axis.text.y = element_text(size = 18),
    axis.title.x = element_text(size = 20),
    plot.title = element_text(size = 24, hjust = 0.5, face = "bold"),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 18),
    strip.text = element_text(size = 18),
    plot.margin = unit(c(1, 1, 1, 2), "lines") 
  ) +
  ggtitle("Upregulated pathways in the SNpc") +
  xlab("Gene Ratio") + ylab("") +
  guides(size = guide_legend(title = "-log10(FDR)"), color = guide_legend(title = "-log10(FDR)")) +
  scale_y_discrete(labels = label_wrap_gen(55)) +
  facet_grid(.~assay)

SN_GO_down_plot <- ggplot(SN_GO_down,
       aes(x = GeneRatioNumeric, y = reorder(Description, -log10(p.adjust)))) +
  geom_point(aes(size = -log10(p.adjust), color = -log10(p.adjust)), alpha = 0.7) + 
  scale_color_gradient(low = "#b7baeb", high = "#5b61b5") + 
  scale_size(range = c(1, 10)) + 
  theme_minimal() + 
  theme(
    axis.text.y = element_text(size = 18),
    axis.title.x = element_text(size = 20),
    plot.title = element_text(size = 24, hjust = 0.5, face = "bold"),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 18),
    strip.text = element_text(size = 18),
    plot.margin = unit(c(1, 1, 1, 5), "lines") 
  ) +
  ggtitle("Downregulated pathways in the SNpc") +
  xlab("Gene Ratio") + ylab("") +
  guides(size = guide_legend(title = "-log10(FDR)"), color = guide_legend(title = "-log10(FDR)")) +
  scale_y_discrete(labels = label_wrap_gen(45)) +
  facet_grid(.~assay)
```

```{r CTX compare ChIP GO to RNA GO}
#filter DEG GO by the top 10 DAR GO
CTX_GO_compare <- CTX_GO_DEG_all %>%
  filter(ID %in% CTX_GO_DAR_top$ID)

CTX_GO_compare$GeneRatioNumeric <- sapply(CTX_GO_compare$GeneRatio, function(x) eval(parse(text = x)))

#filter by most significant 
CTX_GO_compare <- CTX_GO_compare %>%
  group_by(ID) %>%
  slice_min(order_by = p.adjust, n = 1) %>%
  ungroup()

CTX_GO_compare$assay <- "RNA"
CTX_GO_DAR_top$assay <- "ChIP"

CTX_GO <- rbind(CTX_GO_DAR_top,CTX_GO_compare)

CTX_GO <- CTX_GO[,c(2,6,10,12,13)]

CTX_GO <- CTX_GO %>%
  mutate(direction = case_when(
    direction == "Hyper" ~ "Up",
    direction == "Hypo" ~ "Down",
    direction == "UP" ~ "Up",
    direction == "DOWN" ~ "Down",
    TRUE ~ direction))

CTX_GO_up <- CTX_GO[CTX_GO$direction =="Up",]
CTX_GO_down <- CTX_GO[CTX_GO$direction =="Down",]

CTX_GO_up_plot <- ggplot(CTX_GO_up,
       aes(x = GeneRatioNumeric, y = reorder(Description, -log10(p.adjust)))) +
  geom_point(aes(size = -log10(p.adjust), color = -log10(p.adjust)), alpha = 0.7) + 
  scale_color_gradient(low = "#f0c560", high = "#e37c07") + 
  scale_size(range = c(1, 10)) + 
  theme_minimal() + 
  theme(
    axis.text.y = element_text(size = 18),
    axis.title.x = element_text(size = 20),
    plot.title = element_text(size = 24, hjust = 0.5, face = "bold"),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 18),
    strip.text = element_text(size = 18),
    plot.margin = unit(c(1, 1, 1, 5), "lines") 
  ) +
  ggtitle("Upregulated pathways in the Cortex") +
  xlab("Gene Ratio") + ylab("") +
  guides(size = guide_legend(title = "-log10(FDR)"), color = guide_legend(title = "-log10(FDR)")) +
  scale_y_discrete(labels = label_wrap_gen(45)) +
  facet_grid(.~assay)

CTX_GO_down_plot <- ggplot(CTX_GO_down,
       aes(x = GeneRatioNumeric, y = reorder(Description, -log10(p.adjust)))) +
  geom_point(aes(size = -log10(p.adjust), color = -log10(p.adjust)), alpha = 0.7) + 
  scale_color_gradient(low = "#b7baeb", high = "#5b61b5") + 
  scale_size(range = c(1, 10)) + 
  theme_minimal() + 
  theme(
    axis.text.y = element_text(size = 18),
    axis.title.x = element_text(size = 20),
    plot.title = element_text(size = 24, hjust = 0.5, face = "bold"),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 18),
    strip.text = element_text(size = 18),
    plot.margin = unit(c(1, 1, 1, 2), "lines") 
  ) +
  ggtitle("Downregulated pathways in the Cortex") +
  xlab("Gene Ratio") + 
  ylab("") +
  guides(size = guide_legend(title = "-log10(FDR)"), color = guide_legend(title = "-log10(FDR)")) +
  scale_y_discrete(labels = label_wrap_gen(width = 45)) +  
  facet_grid(.~assay)
```

# Figure 3

```{r volcano plot for RNA cortex and SNpc}
## Volcano plot of H3K27ac fold change in rotenone vs control group ##
# get top 20 genes for labelling 
top_genes <- head(DGE_CTX, 20)

# need to add labels for the top 20 genes
CTX_volcano <- ggplot(DGE_CTX, 
                         aes(x=logFC, y=-log10(FDR), colour=DEG)) + 
  geom_point(size=1) + 
  xlim(-5,5) + 
  scale_colour_manual(values=c("#de9483", "#de9483", "#e0b1a6")) + 
  theme_classic() + 
  ylab("-log10(FDR)") + 
  guides(colour="none") + 
  theme(text = element_text(size=24), plot.title = element_text(hjust = 0.5)) + 
  ggtitle("Cortex") + 
  geom_text_repel(data = top_genes, aes(label = Gene), box.padding = 0.5, colour="#506777FF", size=5) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "slategrey") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "slategrey")

CTX_volcano

## Volcano plot of H3K27ac fold change in rotenone vs control group ##
# get top 20 genes for labelling 

top_genes_up <- DGE_SN %>%
  filter(DEG == "UP") %>%
  slice_min(FDR, n = 10)  


top_genes_down <- DGE_SN %>%
  filter(DEG == "DOWN") %>%
  slice_min(FDR, n = 10)  

# Combine the top 10 UP and top 10 DOWN genes
top_genes <- bind_rows(top_genes_up, top_genes_down)
top_genes <- top_genes %>%
  filter(!str_detect(Gene, "LOC"))


# need to add labels for the top 20 genes
SN_volcano <- ggplot(DGE_SN, 
                         aes(x=logFC, y=-log10(FDR), colour=DEG)) + 
  geom_point(size=1) + 
  xlim(-7,7) + 
  scale_colour_manual(values=c("#83d4b9", "#83d4b9", "#afe3d2")) + 
  theme_classic() + 
  ylab("-log10(FDR)") + 
  guides(colour="none") + 
  theme(text = element_text(size=24), plot.title = element_text(hjust = 0.5)) + 
  ggtitle("SN") + 
  geom_text_repel(data = top_genes, aes(label = Gene), box.padding = 0.5, colour="#506777FF", size=5) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "slategrey") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "slategrey")

SN_volcano
```

# Supplementary Figure 1

```{r correlation ChIP promoters with gene expression}
# correlate with ChIP-seq promoter peaks
#SN
SN_promoter_peaks <- SN_DAR_annot[grep("Promoter", SN_DAR_annot$annotation), ]

#get only significant peaks
SN_promoter_peaks <- SN_promoter_peaks[SN_promoter_peaks$FDR < 0.05,]

# get median logFC for each comparison
get_median_logfc <- function(SN_promoter_peaks) {
  SN_promoter_peaks %>%
  group_by(GENE) %>%
  dplyr::summarise(median_log2FoldChange = median(logFC, na.rm = TRUE))
}

# SN rot vs veh
SN_chip_logFC <- get_median_logfc(SN_promoter_peaks)

SN_cor <- merge(DGE_SN, SN_chip_logFC, by.x="Gene", by.y="GENE")

# correlate with ChIP-seq promoter peaks
#Cortex
CTX_promoter_peaks <- CTX_DAR_annot[grep("Promoter", CTX_DAR_annot$annotation), ]

#get only significant peaks
CTX_promoter_peaks <- CTX_promoter_peaks[CTX_promoter_peaks$FDR < 0.05,]

# get median logFC for each comparison
get_median_logfc <- function(CTX_promoter_peaks) {
  CTX_promoter_peaks %>%
  group_by(GENE) %>%
  dplyr::summarise(median_log2FoldChange = median(logFC, na.rm = TRUE))
}

# CTX rot vs veh
CTX_chip_logFC <- get_median_logfc(CTX_promoter_peaks)

CTX_cor <- merge(DGE_CTX, CTX_chip_logFC, by.x="Gene", by.y="GENE")

SN_cor_plot <- ggplot(SN_cor, aes(x = logFC, y = median_log2FoldChange)) +
  geom_point(colour="#83d4b9", size=1) +
  geom_smooth(method = "lm", se = FALSE, colour="#83d4b9", alpha=0.1) +  # Add a linear regression line
  geom_abline(intercept = 0, slope = 0, linetype="dashed", color="black") +  # Add a horizontal line at y=0
  geom_vline(xintercept = 0, linetype="dashed", color="black") +  # Add a vertical line at x=0
  theme_minimal() +  # Set a minimal theme with no background
  labs(title = "SN",
       x = "Expression logFC",
       y = "Acetylation logFC") + 
  stat_cor(size=4) + xlim(-6,6) + ylim(-3,3) + 
  theme(text = element_text(size=16))

CTX_cor_plot <- ggplot(CTX_cor, aes(x = logFC, y = median_log2FoldChange)) +
  geom_point(colour="#de9483", size=1) +
  geom_smooth(method = "lm", se = FALSE, colour="#de9483", alpha=0.1) +  # Add a linear regression line
  geom_abline(intercept = 0, slope = 0, linetype="dashed", color="black") +  # Add a horizontal line at y=0
  geom_vline(xintercept = 0, linetype="dashed", color="black") +  # Add a vertical line at x=0
  theme_minimal() +  # Set a minimal theme with no background
  labs(title = "Cortex",
       x = "Expression logFC",
       y = "Acetylation logFC") + 
  stat_cor(size=4) + xlim(-6,6) + ylim(-3,3) + 
  theme(text = element_text(size=16))
grid.arrange(SN_cor_plot, CTX_cor_plot)
```

```{r save correlation plot}
png(file = "/Users/mariatsalenchuk/Desktop/Bulk_Rat/Manuscript/figures/correlation.png", width = 1000, height = 1000, res = 150)

ggarrange(CTX_cor_plot, SN_cor_plot, common.legend = TRUE, legend = "bottom", nrow = 2, ncol = 1)
dev.off()
```
